lib/cloud-metadata/index.js

```diff
@@ -0,0 +1,197 @@
+'use strict'
+const { getMetadataAwsV1, getMetadataAwsV2 } = require('./aws')
+const { getMetadataAzure } = require('./azure')
+const { getMetadataGcp } = require('./gcp')
+const { CallbackCoordination } = require('./callback-coordination')
+const DEFAULT_PROVIDER_CONFIG = {
+  aws: {
+    host: '169.254.169.254',
+    port: 80,
+    protocol: 'http'
+  },
+  azure: {
+    host: '169.254.169.254',
+    port: 80,
+    protocol: 'http'
+  },
+  gcp: {
+    host: 'metadata.google.internal',
+    port: 80,
+    protocol: 'http'
+  }
+}
+
+// we "ping" (in the colloquial sense, not the ICMP sense) the metadata
+// servers by having a listener that expects the underlying socket to
+// connect.  PING_TIMEOUT_MS control that time
+const PING_TIMEOUT_MS = 100
+
+// some of the metadata servers have a dedicated domain name.  This
+// value is added to the above socket ping to allow extra time for
+// the DNS to resolve
+const DNS_TIMEOUT_MS = 100
+
+// the metadata servers are HTTP services.  This value is
+// used as the timeout for the HTTP request that's
+// made.
+const HTTP_TIMEOUT_MS = 1000
+
+// timeout for the CallbackCoordination object -- this is a fallback to
+// account for a catastrophic error in the CallbackCoordination object
+const COORDINATION_TIMEOUT_MS = 3000
+
+class CloudMetadata {
+  constructor (agent) {
+    this.agent = agent
+  }
+
+  /**
+   * Fetches Cloud Metadata
+   *
+   * The module's main entry-point/primary method.  The getCloudMetadata function
+   * will query the cloud metadata servers and return (via a callback function)
+   * to final metadata object.  This function may be called with a single
+   * argument.
+   *
+   * getCloudMetadata(function(error, metadata){
+   *     //...
+   * })
+   *
+   * Or with two
+   *
+   * getCloudMetadata(config, function(error, metadata){
+   *     //...
+   * })
+   *
+   * The config parameter is an object that contains information on the
+   * metadata servers.  If omitted, the function will use the  module-global
+   * DEFAULT_PROVIDER_CONFIG values
+   */
+  getCloudMetadata (config, cb) {
+    // normalize arguments
+    if (!cb) {
+      cb = config
+      config = DEFAULT_PROVIDER_CONFIG
+    }
+    const fetcher = new CallbackCoordination(COORDINATION_TIMEOUT_MS)
+
+    if (this.shouldFetchGcp()) {
+      fetcher.schedule(function (fetcher) {
+        getMetadataGcp(
+          config.gcp.host,
+          config.gcp.port,
+          PING_TIMEOUT_MS + DNS_TIMEOUT_MS,
+          HTTP_TIMEOUT_MS,
+          config.gcp.protocol,
+          (error, result) => {
+            fetcher.recordResult(error, result)
+          }
+        )
+      })
+    }
+
+    if (this.shouldFetchAws()) {
+      // amazon has two metadata endpoints -- IMDSv1 and IMDSv2
+      // we schedule a callback to attempt to fetch from each
+      // one -- which callback succeeds first will win
```

Comment:
> My (possibly naive) read of the IMDS docs is that IMDSv2 should work everywhere now (i.e. there shouldn't be a case of a customer running on an "old" ECS VM whose metadata endpoint doesn't support v2).

Regardless of what choice we make here, I don't think we can assume that IMDSv2 is available "everywhere".  Putting aside the case of an old EC2 VM that's been up since before the V2 endpoint was rolled out in [Nov '19](https://aws.amazon.com/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/) (it sounds banana-cakes but I've seen things) -- there's just too many other AWS services that _might_ run node code where the V2 metadata service isn't available. 

For example, I know EKS only turned the V2 endpoint on in August of this year: https://aws.amazon.com/about-aws/whats-new/2020/08/amazon-eks-supports-ec2-instance-metadata-service-v2/ -- I don't know what other services out there might support one or the other, and figuring that out is one of those boil the ocean problems that might not be the best use of our time. 

Re: extra perf. -- it's good to think through the performance tradeoffs, but this isn't a hot path -- it's on agent startup  and it happens once. The amount of extra work is, in the grand scheme of things, small. Also, as you noted, with this implementation if the V1 endpoint _does_ return data earlier startup proceeds slightly faster -- but from my point of view perf. is basically a wash one way or the other.

All that said 

> showing up in CloudWatch metrics for those following https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html#instance-metadata-transition-to-version-2 to try to migrate fully away from v1 for security

That seems a legitimate reason not to make both calls.  We still need to handle the case where the IMDSv1 endpoint is the only one -- so I'll take python's approach of only setting the header when the token server actually responds.  If that token endpoint fails to respond, we'll call the V1 endpoint.

Reply:
Nice find with the EKS case. Your approach sounds good to me.